build.gradle.kts

```diff
@@ -188,78 +187,113 @@ nexusPublishing {
 	}
 }
 
-tasks {
+// Our Task to call all our versionTests
+val versionTest by tasks.creating
 
-	sourceSets {
-		main {
-			if (modularBuild.toBoolean())
-				java.srcDir("src/main/module")
-		}
-		create("demo") {
-			java {
-				srcDir("src/demo/java")
+testing {
+	suites {
+		val test by getting(JvmTestSuite::class) {
+			useJUnitJupiter()
+
+			targets {
+				all {
+					testTask.configure {
+						configure<JacocoTaskExtension> {
+							isEnabled = !experimentalBuild
+						}
+						testLogging {
+							setExceptionFormat("full")
+						}
+						useJUnitPlatform()
+						filter {
+							includeTestsMatching("*Tests")
+						}
+						systemProperty("java.util.logging.manager", "org.apache.logging.log4j.jul.LogManager")
+						// `EnvironmentVariableExtension` uses reflection to change environment variables;
+						// this prevents the corresponding warning (and keeps working on Java 8)
+						// IF YOU ADD MORE OPTIONS; CONSIDER REPLACING `-XX:+IgnoreUnrecognizedVMOptions WITH A CONDITIONAL
+						jvmArgs(
+								"-XX:+IgnoreUnrecognizedVMOptions",
+								"--add-opens=java.base/java.util=ALL-UNNAMED")
+
+					}
+				}
 			}
-			compileClasspath += sourceSets.main.get().output
-			runtimeClasspath += sourceSets.main.get().output
-		}
-	}
-	project(":demo") {
-		sonarqube {
-			isSkipProject = true
 		}
-	}
-	// Adds all dependencies of main to demo sourceSet
-	configurations["demoImplementation"].extendsFrom(configurations.implementation.get())
-	// Ensures JUnit 5 engine is available to demo at runtime
-	configurations["demoRuntimeOnly"].extendsFrom(configurations.testImplementation.get())
 
-	compileJava {
-		options.encoding = "UTF-8"
-	}
-
-	compileTestJava {
-		options.encoding = "UTF-8"
-	}
+		val demoTests by registering(JvmTestSuite::class) {
+			dependencies { implementation(project) }
 
-	test {
-		configure<JacocoTaskExtension> {
-			isEnabled = !experimentalBuild
-		}
-		testLogging {
-			setExceptionFormat("full")
-		}
-		useJUnitPlatform()
-		filter {
-			includeTestsMatching("*Tests")
+			sources { java { srcDir("src/demo/java") } }
+			targets {
+				all {
+					testTask.configure {
+						shouldRunAfter(test)
+						filter {
+							includeTestsMatching("*Demo")
+						}
+					}
+				}
+			}
 		}
-		systemProperty("java.util.logging.manager", "org.apache.logging.log4j.jul.LogManager")
-		// `EnvironmentVariableExtension` uses reflection to change environment variables;
-		// this prevents the corresponding warning (and keeps working on Java 8)
-		// IF YOU ADD MORE OPTIONS; CONSIDER REPLACING `-XX:+IgnoreUnrecognizedVMOptions WITH A CONDITIONAL
-		jvmArgs(
-				"-XX:+IgnoreUnrecognizedVMOptions",
-				"--add-opens=java.base/java.util=ALL-UNNAMED")
-	}
 
-	testing {
-		suites {
-			val test by getting(JvmTestSuite::class) {
+		supportedJUnitVersions.split(",").filter { it != junitVersion }.forEach {
```

Comment:
it is a thing - but  there might be a nicer version.

Generally speaking, this reduced the CI minutes - we only have 1 job to prep and boot up - instead of one per junit verison. the whole orchestration is removed, and we are down to just a simple additional test execution. So in the end it makes our builds faster.

Additionally i am a fan of, having less configuration in CI and make it easy for people to use ci like feature locally - with this, we have an own gradle target to test multiple versions at once. So i do think this makes the local development process easier and nice

Reply:
> with this, we have an own gradle target to test multiple versions at once. So i do think this makes the local development process easier and nice

Sounds to me like: You can test multiple versions locally, but our CI does not anymore.
This would further mean that we force the users to run their changes locally instead of seeing impacts on Github and decide there what (not) to do?!?